tourist_scheduling_system/src/agents/scheduler_agent.py

```diff
@@ -0,0 +1,412 @@
+2. Receives GuideOffers from guide agents
+3. Runs greedy scheduling algorithm to match tourists to guides
+4. Sends ScheduleProposals back to requesting tourists
+
+This is implemented as an A2A server using the official a2a-sdk.
+"""
+
+import json
+import logging
+import time
+from pathlib import Path
+from dataclasses import dataclass
+from datetime import datetime
+
+import click
+import uvicorn
+import requests
+from a2a.server.agent_execution import AgentExecutor, RequestContext
+from a2a.server.apps import A2AStarletteApplication
+from a2a.server.events import EventQueue
+from a2a.server.request_handlers import DefaultRequestHandler
+from a2a.server.tasks import InMemoryTaskStore
+from a2a.server.tasks.task_updater import TaskUpdater
+from a2a.utils.task import new_task
+from a2a.types import (
+    AgentCapabilities,
+    AgentCard,
+    AgentSkill,
+    Part,
+    TaskState,
+    TextPart,
+    DataPart,  # Allows structured JSON parts without manual decoding
+)
+
+from core.messages import (
+    Assignment,
+    GuideOffer,
+    ScheduleProposal,
+    TouristRequest,
+    Window,
+)
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+# In-memory storage for scheduler state
+@dataclass
+class SchedulerState:
+    """Centralized scheduler state"""
+
+    tourist_requests: list[TouristRequest]
+    guide_offers: list[GuideOffer]
+    assignments: list[Assignment]
+
+    def __init__(self):
+        self.tourist_requests = []
+        self.guide_offers = []
+        self.assignments = []
+
+
+state = SchedulerState()
+
+
+def _discover_ui_ports() -> int:
+    """Attempt to discover the UI agent A2A port from written ports file.
+
+    Returns discovered port or default 10012. Non-fatal if file missing.
+    """
+    default_port = 10012
+    try:
+        ports_file = Path(__file__).resolve().parent.parent / "ui_agent_ports.json"
+        if ports_file.exists():
+            data = json.loads(ports_file.read_text())
+            a2a_port = int(data.get("a2a_port", default_port))
+            return a2a_port
+    except Exception as e:
+        logger.debug(f"[Scheduler] UI ports file read failed: {e}")
+    return default_port
+
+async def send_to_ui_agent(message_data: dict):
+    """Send data to UI agent for dashboard updates, using dynamic port discovery."""
+    port = _discover_ui_ports()
+    url = f"http://localhost:{port}/"
+    try:
+        response = requests.post(
+            url,
+            json={
+                "jsonrpc": "2.0",
+                "id": f"scheduler-ui-{int(time.time())}",
+                "method": "message/send",
+                "params": {
+                    "message": {
+                        "role": "user",
+                        "parts": [{"kind": "text", "text": json.dumps(message_data)}],
+                        "messageId": f"scheduler-msg-{int(time.time())}"
+                    }
+                }
+            },
+            timeout=3
+        )
```

Comment:
why is this using raw requests instead of an A2A client?

Should use httpx http client for asyncio support instead of requests otherwise it can block the event thread for the entire time the request is running